## Model evaluation



### Why Use MAE, RMSE, and MAPE?
In forecasting model evaluation, the goal is to measure how well the model's predictions fit the actual data. To achieve this, error metrics are used to quantify the difference between predictions (`yhat`) and observed values (`y`). Three of the most common metrics for this purpose are:

#### 1. **Mean Absolute Error (MAE)**
   - **Definition**: MAE is the average of the absolute errors between the predictions and actual values. It measures the average magnitude of errors in a set of predictions, regardless of whether the errors are positive or negative.
   - **Formula**: 
     $$ \text{MAE} = \frac{1}{n} \sum_{i=1}^n |y_i - \hat{y}_i|$$
     
   - **Interpretation**: MAE provides an intuitive measure of how much our predictions deviate on average from the actual values. Since it does not square the errors, MAE is less sensitive to outliers, making it useful when we want a robust measure of accuracy that isn't heavily skewed by unexpected spikes.
   - **Applicability in COVID-19 forecasting**: If the priority is having a "typical" error that is easy to interpret in terms of daily case numbers, MAE is a good choice.

#### 2. **Root Mean Squared Error (RMSE)**
   - **Definition**: RMSE is the square root of the average of the squared differences between predicted and actual values. Unlike MAE, RMSE penalizes larger errors more heavily, which can be beneficial when large errors represent a higher risk in decision-making.
   - **Formula**:
     $$ \text{RMSE} = \sqrt{\frac{1}{n} \sum_{i=1}^n (y_i - \hat{y}_i)^2}$$
     
   - **Interpretation**: RMSE is more sensitive to outliers, which can be useful if you want to ensure that the model penalizes large errors significantly. In COVID-19 forecasting, RMSE can be particularly useful when large deviations in case numbers are concerning (for example, during peak periods).
   - **Applicability in COVID-19 forecasting**: If large deviations in case numbers are especially problematic, RMSE is a suitable metric as it "punishes" predictions that stray significantly from actual values.

#### 3. **Mean Absolute Percentage Error (MAPE)**
   - **Definition**: MAPE calculates the average of the absolute percentage errors between predictions and actual values, providing an intuitive percentage-based accuracy metric.
   - **Formula**:
     $$ \text{MAPE} = \frac{100}{n} \sum_{i=1}^n \left| \frac{y_i - \hat{y}_i}{y_i} \right|$$
     
   - **Interpretation**: MAPE expresses errors as a percentage of the actual values, making it easy to interpret in terms of relative accuracy. For instance, a MAPE of 5% means the average forecasted value is off by 5% from the actual value. However, MAPE can be sensitive to very small actual values, leading to high error rates or undefined values.
   - **Applicability in COVID-19 forecasting**: MAPE is especially useful when it's important to understand the error in relative terms, such as how much the model’s predictions deviate as a percentage of actual cases. This is beneficial for comparing performance over time or across regions with varying case numbers.

### Why Use All Three Metrics?
MAE, RMSE, and MAPE each provide unique insights into the model’s performance:

- **MAE** gives a straightforward measure of average error, useful for understanding typical deviations in absolute terms and for robust performance when outliers are present.
- **RMSE** is beneficial if we want the model to be "strict" and heavily penalize larger errors, which is valuable in situations where large deviations are particularly costly or concerning.
- **MAPE** allows us to assess the model’s accuracy in percentage terms, which is helpful for comparing accuracy across datasets of different scales or understanding performance across varying levels of COVID-19 cases.

By using all three metrics, you can achieve a more balanced and comprehensive view of the model's performance, capturing both absolute and relative error perspectives, as well as sensitivity to large deviations.

In [1]:
import numpy as np
import pandas as pd
import malib.models.metrics as mmm
import malib.data.plotting as mdp
import seaborn as sns
import matplotlib.pyplot as plt
import plotly.express as px

In [2]:
PATH = "../../data/processed/df_clean_target_confirmed.csv"
PATH_HP_OPTUNA = "../../data/processed/prediction/optuna.csv"
PATH_HP_HYPEROPT = "../../data/processed/prediction/hyperopt.csv"
PATH_HP_PARAM_GRID = "../../data/processed/prediction/params_grid.csv"
PATH_TRAIN = "../../data/processed/train.csv"
PATH_TEST = "../../data/processed/test.csv"

In [3]:
df = pd.read_csv(PATH)
df_opt = pd.read_csv(PATH_HP_OPTUNA)
df_hyp = pd.read_csv(PATH_HP_HYPEROPT)
df_pg = pd.read_csv(PATH_HP_PARAM_GRID)
train = pd.read_csv(PATH_TRAIN)

In [4]:
list_of_models = {"optuna":df_opt,"hyperopt":df_hyp,"params_grid":df_pg}

In [5]:
rr = df_opt.shape[0]

In [6]:
for key,m in list_of_models.items():
    print(key)
    eval = mmm.daily_evaluation(df.tail(rr),m,"ds","y","yhat")
    display(eval)
    print("Metrics summary")
    pd.set_option('display.float_format', lambda x: '%.3f' % x)
    display(eval.mean(numeric_only=True).T)
    eval["y"] = eval["predict"]
    eval["ds"] = eval["Date"]
    fig = mdp.plot_interactive_forecast(df,train,eval[["ds","y"]],"ds","y")
    fig.show(renderer='iframe')

optuna


Date        true      predict        RMSE         MAE      MAPE  \
0   2020-06-20   8805336.0   8720314.41    85021.59    85021.59  0.009656   
1   2020-06-21   8933875.0   8846325.26    87549.74    87549.74  0.009800   
2   2020-06-22   9071733.0   8970091.23   101641.77   101641.77  0.011204   
3   2020-06-23   9237071.0   9099658.22   137412.78   137412.78  0.014876   
4   2020-06-24   9408254.0   9231990.93   176263.07   176263.07  0.018735   
5   2020-06-25   9586141.0   9367322.47   218818.53   218818.53  0.022827   
6   2020-06-26   9777487.0   9502945.01   274541.99   274541.99  0.028079   
7   2020-06-27   9955597.0   9631877.25   323719.75   323719.75  0.032516   
8   2020-06-28  10117227.0   9757888.10   359338.90   359338.90  0.035518   
9   2020-06-29  10275799.0   9881654.07   394144.93   394144.93  0.038357   
10  2020-06-30  10449697.0  10011221.06   438475.94   438475.94  0.041961   
11  2020-07-01  10667386.0  10143553.77   523832.23   523832.23  0.049106   
12  2020-07-02  10875091.0  10278885.31   596205.69   596205.69  0.054823   
13  2020-07-03  11078585.0  10414507.85   664077.15   664077.15  0.059942   
14  2020-07-04  11272152.0  10543440.09   728711.91   728711.91  0.064647   
15  2020-07-05  11454847.0  10669450.94   785396.06   785396.06  0.068565   
16  2020-07-06  11622190.0  10793216.91   828973.09   828973.09  0.071327   
17  2020-07-07  11833034.0  10922783.90   910250.10   910250.10  0.076924   
18  2020-07-08  12044836.0  11055116.61   989719.39   989719.39  0.082170   
19  2020-07-09  12273063.0  11190448.15  1082614.85  1082614.85  0.088211   
20  2020-07-10  12505640.0  11326070.69  1179569.31  1179569.31  0.094323   
21  2020-07-11  12721968.0  11455002.93  1266965.07  1266965.07  0.099589   
22  2020-07-12  12914636.0  11581013.78  1333622.22  1333622.22  0.103264   
23  2020-07-13  13107415.0  11704779.75  1402635.25  1402635.25  0.107011   
24  2020-07-14  13328867.0  11834346.74  1494520.26  1494520.26  0.112127   
25  2020-07-15  13559984.0  11966679.45  1593304.55  1593304.55  0.117500   
26  2020-07-16  13812525.0  12102010.99  1710514.01  1710514.01  0.123838   
27  2020-07-17  14054563.0  12237633.53  1816929.47  1816929.47  0.129277   
28  2020-07-18  14292198.0  12366565.77  1925632.23  1925632.23  0.134733   
29  2020-07-19  14506845.0  12492576.62  2014268.38  2014268.38  0.138850   
30  2020-07-20  14713623.0  12616342.60  2097280.40  2097280.40  0.142540   
31  2020-07-21  14947078.0  12745909.58  2201168.42  2201168.42  0.147264   
32  2020-07-22  15227725.0  12878242.30  2349482.70  2349482.70  0.154290   
33  2020-07-23  15510481.0  13013573.83  2496907.17  2496907.17  0.160982   
34  2020-07-24  15791645.0  13149196.37  2642448.63  2642448.63  0.167332   
35  2020-07-25  16047190.0  13278128.61  2769061.39  2769061.39  0.172557   
36  2020-07-26  16251796.0  13404139.46  2847656.54  2847656.54  0.175221   
37  2020-07-27  16480485.0  13527905.44  2952579.56  2952579.56  0.179156   

    Normalized RMSE  Normalized MAE  Accuracy  
0          0.009656        0.009656  0.990344  
1          0.009800        0.009800  0.990200  
2          0.011204        0.011204  0.988796  
3          0.014876        0.014876  0.985124  
4          0.018735        0.018735  0.981265  
5          0.022827        0.022827  0.977173  
6          0.028079        0.028079  0.971921  
7          0.032516        0.032516  0.967484  
8          0.035518        0.035518  0.964482  
9          0.038357        0.038357  0.961643  
10         0.041961        0.041961  0.958039  
11         0.049106        0.049106  0.950894  
12         0.054823        0.054823  0.945177  
13         0.059942        0.059942  0.940058  
14         0.064647        0.064647  0.935353  
15         0.068565        0.068565  0.931435  
16         0.071327        0.071327  0.928673  
17         0.076924        0.076924  0.923076  
18         0.082170        0.082170  0.917830  
19         0.088211        0.088211  0.911789 

Metrics summary


true              12329317.500
predict           11124021.315
RMSE               1205296.185
MAE                1205296.185
MAPE                     0.088
Normalized RMSE          0.088
Normalized MAE           0.088
Accuracy                 0.912
dtype: float64

/home/marodenas/Documentos/data_science/repositorios/covid_forecast_prophet/src/malib/data/plotting.py:373: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  prediction_data["Series"] = "Predictions"


hyperopt


Date         true      predict        RMSE         MAE  MAPE  \
0   2020-06-20  8805336.000  8720811.280   84524.720   84524.720 0.010   
1   2020-06-21  8933875.000  8845628.310   88246.690   88246.690 0.010   
2   2020-06-22  9071733.000  8968951.410  102781.590  102781.590 0.011   
3   2020-06-23  9237071.000  9098808.930  138262.070  138262.070 0.015   
4   2020-06-24  9408254.000  9232185.310  176068.690  176068.690 0.019   
5   2020-06-25  9586141.000  9368464.730  217676.270  217676.270 0.023   
6   2020-06-26  9777487.000  9504338.900  273148.100  273148.100 0.028   
7   2020-06-27  9955597.000  9632509.380  323087.620  323087.620 0.032   
8   2020-06-28 10117227.000  9757326.420  359900.580  359900.580 0.036   
9   2020-06-29 10275799.000  9880649.510  395149.490  395149.490 0.038   
10  2020-06-30 10449697.000 10010507.030  439189.970  439189.970 0.042   
11  2020-07-01 10667386.000 10143883.410  523502.590  523502.590 0.049   
12  2020-07-02 10875091.000 10280162.840  594928.160  594928.160 0.055   
13  2020-07-03 11078585.000 10416037.000  662548.000  662548.000 0.060   
14  2020-07-04 11272152.000 10544207.480  727944.520  727944.520 0.065   
15  2020-07-05 11454847.000 10669024.520  785822.480  785822.480 0.069   
16  2020-07-06 11622190.000 10792347.620  829842.380  829842.380 0.071   
17  2020-07-07 11833034.000 10922205.140  910828.860  910828.860 0.077   
18  2020-07-08 12044836.000 11055581.520  989254.480  989254.480 0.082   
19  2020-07-09 12273063.000 11191860.940 1081202.060 1081202.060 0.088   
20  2020-07-10 12505640.000 11327735.110 1177904.890 1177904.890 0.094   
21  2020-07-11 12721968.000 11455905.590 1266062.410 1266062.410 0.100   
22  2020-07-12 12914636.000 11580722.630 1333913.370 1333913.370 0.103   
23  2020-07-13 13107415.000 11704045.720 1403369.280 1403369.280 0.107   
24  2020-07-14 13328867.000 11833903.240 1494963.760 1494963.760 0.112   
25  2020-07-15 13559984.000 11967279.620 1592704.380 1592704.380 0.117   
26  2020-07-16 13812525.000 12103559.050 1708965.950 1708965.950 0.124   
27  2020-07-17 14054563.000 12239433.210 1815129.790 1815129.790 0.129   
28  2020-07-18 14292198.000 12367603.690 1924594.310 1924594.310 0.135   
29  2020-07-19 14506845.000 12492420.730 2014424.270 2014424.270 0.139   
30  2020-07-20 14713623.000 12615743.830 2097879.170 2097879.170 0.143   
31  2020-07-21 14947078.000 12745601.350 2201476.650 2201476.650 0.147   
32  2020-07-22 15227725.000 12878977.720 2348747.280 2348747.280 0.154   
33  2020-07-23 15510481.000 13015257.150 2495223.850 2495223.850 0.161   
34  2020-07-24 15791645.000 13151131.320 2640513.680 2640513.680 0.167   
35  2020-07-25 16047190.000 13279301.800 2767888.200 2767888.200 0.172   
36  2020-07-26 16251796.000 13404118.840 2847677.160 2847677.160 0.175   
37  2020-07-27 16480485.000 13527441.930 2953043.070 2953043.070 0.179   

    Normalized RMSE  Normalized MAE  Accuracy  
0             0.010           0.010     0.990  
1             0.010           0.010     0.990  
2             0.011           0.011     0.989  
3             0.015           0.015     0.985  
4             0.019           0.019     0.981  
5             0.023           0.023     0.977  
6             0.028           0.028     0.972  
7             0.032           0.032     0.968  
8             0.036           0.036     0.964  
9             0.038           0.038     0.962  
10            0.042           0.042     0.958  
11            0.049           0.049     0.951  
12            0.055           0.055     0.945  
13            0.060           0.060     0.940  
14            0.065           0.065     0.935  
15            0.069           0.069     0.931  
16            0.071           0.071     0.929  
17            0.077           0.077     0.923  
18            0.082           0.082     0.918  
19            0.088           0.088     0.912  
20            0.094           0.094     0.906  
21            0.100           0.100     0.900  
22            0.103

Metrics summary


true              12329317.500
predict           11124359.848
RMSE               1204957.652
MAE                1204957.652
MAPE                     0.088
Normalized RMSE          0.088
Normalized MAE           0.088
Accuracy                 0.912
dtype: float64

/home/marodenas/Documentos/data_science/repositorios/covid_forecast_prophet/src/malib/data/plotting.py:373: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



params_grid


Date         true      predict        RMSE         MAE  MAPE  \
0   2020-06-20  8805336.000  8725347.620   79988.380   79988.380 0.009   
1   2020-06-21  8933875.000  8851206.460   82668.540   82668.540 0.009   
2   2020-06-22  9071733.000  8974724.740   97008.260   97008.260 0.011   
3   2020-06-23  9237071.000  9103950.770  133120.230  133120.230 0.014   
4   2020-06-24  9408254.000  9237104.000  171150.000  171150.000 0.018   
5   2020-06-25  9586141.000  9374902.960  211238.040  211238.040 0.022   
6   2020-06-26  9777487.000  9511477.830  266009.170  266009.170 0.027   
7   2020-06-27  9955597.000  9641308.450  314288.550  314288.550 0.032   
8   2020-06-28 10117227.000  9767167.290  350059.710  350059.710 0.035   
9   2020-06-29 10275799.000  9890685.560  385113.440  385113.440 0.037   
10  2020-06-30 10449697.000 10019911.590  429785.410  429785.410 0.041   
11  2020-07-01 10667386.000 10153064.830  514321.170  514321.170 0.048   
12  2020-07-02 10875091.000 10290863.790  584227.210  584227.210 0.054   
13  2020-07-03 11078585.000 10427438.650  651146.350  651146.350 0.059   
14  2020-07-04 11272152.000 10557269.270  714882.730  714882.730 0.063   
15  2020-07-05 11454847.000 10683128.110  771718.890  771718.890 0.067   
16  2020-07-06 11622190.000 10806646.390  815543.610  815543.610 0.070   
17  2020-07-07 11833034.000 10935872.420  897161.580  897161.580 0.076   
18  2020-07-08 12044836.000 11069025.650  975810.350  975810.350 0.081   
19  2020-07-09 12273063.000 11206824.610 1066238.390 1066238.390 0.087   
20  2020-07-10 12505640.000 11343399.480 1162240.520 1162240.520 0.093   
21  2020-07-11 12721968.000 11473230.100 1248737.900 1248737.900 0.098   
22  2020-07-12 12914636.000 11599088.940 1315547.060 1315547.060 0.102   
23  2020-07-13 13107415.000 11722607.220 1384807.780 1384807.780 0.106   
24  2020-07-14 13328867.000 11851833.250 1477033.750 1477033.750 0.111   
25  2020-07-15 13559984.000 11984986.480 1574997.520 1574997.520 0.116   
26  2020-07-16 13812525.000 12122785.440 1689739.560 1689739.560 0.122   
27  2020-07-17 14054563.000 12259360.310 1795202.690 1795202.690 0.128   
28  2020-07-18 14292198.000 12389190.930 1903007.070 1903007.070 0.133   
29  2020-07-19 14506845.000 12515049.770 1991795.230 1991795.230 0.137   
30  2020-07-20 14713623.000 12638568.040 2075054.960 2075054.960 0.141   
31  2020-07-21 14947078.000 12767794.070 2179283.930 2179283.930 0.146   
32  2020-07-22 15227725.000 12900947.310 2326777.690 2326777.690 0.153   
33  2020-07-23 15510481.000 13038746.270 2471734.730 2471734.730 0.159   
34  2020-07-24 15791645.000 13175321.140 2616323.860 2616323.860 0.166   
35  2020-07-25 16047190.000 13305151.760 2742038.240 2742038.240 0.171   
36  2020-07-26 16251796.000 13431010.600 2820785.400 2820785.400 0.174   
37  2020-07-27 16480485.000 13554528.870 2925956.130 2925956.130 0.178   

    Normalized RMSE  Normalized MAE  Accuracy  
0             0.009           0.009     0.991  
1             0.009           0.009     0.991  
2             0.011           0.011     0.989  
3             0.014           0.014     0.986  
4             0.018           0.018     0.982  
5             0.022           0.022     0.978  
6             0.027           0.027     0.973  
7             0.032           0.032     0.968  
8             0.035           0.035     0.965  
9             0.037           0.037     0.963  
10            0.041           0.041     0.959  
11            0.048           0.048     0.952  
12            0.054           0.054     0.946  
13            0.059           0.059     0.941  
14            0.063           0.063     0.937  
15            0.067           0.067     0.933  
16            0.070           0.070     0.930  
17            0.076           0.076     0.924  
18            0.081           0.081     0.919  
19            0.087           0.087     0.913  
20            0.093           0.093     0.907  
21            0.098           0.098     0.902  
22            0.102

Metrics summary


true              12329317.500
predict           11139513.710
RMSE               1189803.790
MAE                1189803.790
MAPE                     0.087
Normalized RMSE          0.087
Normalized MAE           0.087
Accuracy                 0.913
dtype: float64

/home/marodenas/Documentos/data_science/repositorios/covid_forecast_prophet/src/malib/data/plotting.py:373: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



### Conclusions

#### 1. Choice of Hyperparameter Tuning Method: Optuna, Param Grid, or Hyperopt

When comparing the performance metrics across Optuna, Param Grid, and Hyperopt, we observe the following:

- **Accuracy**: All three methods achieve a very similar average accuracy of approximately 0.912–0.913. This suggests that each tuning method is producing models with a comparable level of overall fit. However, **Param Grid** has a slightly higher accuracy (0.913) than the others, though the difference is minimal.
  
- **RMSE and MAE**: Param Grid achieves the lowest overall RMSE (1,189,803) and MAE (1,189,803) among the three methods. Hyperopt follows closely, while Optuna has a slightly higher RMSE and MAE. This indicates that **Param Grid** provides the best absolute fit in terms of error, as both RMSE and MAE are minimized with this method.

- **MAPE and Normalized RMSE/MAE**: The average MAPE and normalized RMSE/MAE for all three methods are also very similar (around 0.087–0.088). However, **Param Grid** has a marginally better normalized RMSE and MAE (0.087) compared to the other methods (0.088 for Optuna and Hyperopt).

**Recommendation**: Based on these metrics, **Param Grid** is the preferred hyperparameter tuning method. It provides the lowest error (RMSE and MAE) and slightly higher accuracy, indicating a marginally better fit. This advantage, though subtle, makes Param Grid the most effective method for hyperparameter tuning in this case.

---

#### 2. Evaluation of Prophet’s Forecasting Performance

The evaluation metrics give insights into the quality of the Prophet model’s forecast:

- **Overall Fit**: With a **MAPE around 8.7%–8.8%**, the Prophet model shows reasonable accuracy, though it suggests there is still some level of deviation from the actual data. A MAPE below 10% is often considered a sign of a fairly good fit in time series forecasting, indicating that Prophet is generally capturing the trend.

- **RMSE and MAE Values**: The RMSE values around 1,200,000 indicate an average forecast deviation in the range of 1.2 million units per prediction. Although these errors are sizable, they may be expected in high-scale time series data (like COVID-19 case numbers) where the absolute values are high. The **normalized RMSE and MAE (~8.7%)** imply that the error is roughly 8.7% of the mean true values, which provides a clearer interpretation of the model’s performance relative to the data scale.

- **Accuracy**: The accuracy values close to 91.3% across the models indicate that Prophet provides a consistent fit across different days, achieving reliable performance over the forecast horizon. However, the daily metrics reveal that the error increases slightly as the forecast horizon extends, which is typical in time series forecasting due to the cumulative uncertainty in the data.

**Overall Assessment**: The Prophet model is performing reasonably well in capturing the overall trend, achieving an average error (MAPE) of around 8.7% and an accuracy close to 91.3%. This indicates a fairly good fit to the observed data. However, there is room for improvement if higher accuracy is required, particularly for reducing absolute error (RMSE) on larger-scale predictions. Further fine-tuning or considering additional seasonal adjustments or other modeling approaches could potentially enhance forecast accuracy for more precise applications.